### ISSUE TITLE
grade predictor finalized

### ISSUE THREAD

**Issue by tarunrayavarapu** [luojonah/palomarhealth_frontend#33]
###  Overview

Our **AP Grade Predictor** is a machine learning-powered web feature designed to give students an estimate of their grade based on their self-assessment. This project is integrated into the Palomar Health Frontend and leverages a trained regression model to generate both a percentage score and a letter grade. The results are saved to the user’s profile for easy retrieval.

---

### The AI Model

We use a **Linear Regression** model trained on a custom dataset (`ap_predict_data.csv`) that includes 11 student-related input features:

- Attendance  
- Work Habits  
- Behavior  
- Timeliness  
- Advocacy  
- Tech Growth  
- Tech Sense  
- Tech Talk  
- Communication and Collaboration  
- Leadership  
- Integrity

Grades are mapped to numerical percentages using the following logic:

Before prediction, each input value is **binarized**:
- Values 1–3 become `0`
- Values 4–5 become `1`

#### Prediction Output

The model predicts:
- A **percentage score** (0–100 range, clamped)
- A **letter grade** based on this scale:
  - 90+ → A  
  - 80–89 → B  
  - 70–79 → C  
  - 60–69 → D  
  - Below 60 → F

---

### Backend API Implementation

We use a Flask Blueprint and `flask_restful` to serve the AI model with JWT protection for user-based history.

#### `grade_model.py`

```python
import pandas as pd
from sklearn.linear_model import LinearRegression

class GradePredictionModel:
    def __init__(self):
        # Load dataset
        data = pd.read_csv("datasets/ap_predict_data.csv")
        data = data.dropna(subset=['Grade'])
        data.columns = data.columns.str.strip()

        # Map grades to numeric percentages
        grade_map = {'A': 90, 'B': 80, 'C': 70, 'D': 60, 'F': 50}
        data['GradePercent'] = data['Grade'].map(grade_map)

        self.features = ['Attendance','Work Habits','Behavior','Timeliness','Advocacy','Tech Growth',
                         'Tech Sense','Tech Talk','Communication and Collaboration','Leadership','Integrity']

        X = data[self.features]
        y = data['GradePercent']

        self.model = LinearRegression()
        self.model.fit(X, y)

    def predict(self, user_input):
        if len(user_input) != len(self.features):
            raise ValueError(f"Expected {len(self.features)} inputs, got {len(user_input)}")

        binarized = [0 if x <= 3 else 1 for x in user_input]

        percent = self.model.predict([binarized])[0]
        percent = max(0, min(100, percent))

        if percent >= 90:
            letter = 'A'
        elif percent >= 80:
            letter = 'B'
        elif percent >= 70:
            letter = 'C'
        elif percent >= 60:
            letter = 'D'
        else:
            letter = 'F'

        return round(percent, 2), letter


---

from flask import Blueprint, request, jsonify, g
from flask_restful import Api, Resource
from api.jwt_authorize import token_required
from model.user import User
from model.grade_model import GradePredictionModel

grade_api = Blueprint('grade_api', __name__, url_prefix='/api/grade')
api = Api(grade_api)

model_instance = GradePredictionModel()

class Predict(Resource):
    def post(self):
        data = request.get_json()
        if not data or 'inputs' not in data:
            return {"error": "Missing 'inputs' field in JSON payload. Expected a list of 11 numbers (1-5)."}, 400

        user_input = data['inputs']

        if len(user_input) != 11:
            return {"error": f"Expected 11 input values, received {len(user_input)}."}, 400

        try:
            user_input = [int(val) for val in user_input]
        except ValueError:
            return {"error": "All input values must be numbers between 1 and 5."}, 400

        if not all(1 <= val <= 5 for val in user_input):
            return {"error": "Input values should be between 1 and 5."}, 400

        percent, letter = model_instance.predict(user_input)

        return jsonify({
            'predicted_percent': percent,
            'predicted_grade': letter
        })

    @token_required()
    def get(self):
        user = g.current_user
        return jsonify(user.grade_data)

api.add_resource(Predict, '/predict')

---


---

# AP Computer Science Principles Exam Predictor
## Technical Documentation & User Stories

---

## 📋 Project Overview

The AP Computer Science Principles Exam Predictor is a comprehensive web application designed to help students estimate their potential AP exam scores based on practice test performance and manual score input. The application features two prediction modes: a practice-based tracker that analyzes historical performance data, and a manual calculator for immediate score estimation.

---

## Core Features

### 1. Practice Exam Tracker
- **Multi-year Practice Tests**: Supports 2018, 2020, and 2021 MCQ practice exams
- **Performance Task Scoring**: Includes practice FRQ (Free Response Questions) evaluation
- **Time Tracking**: Records time spent on each practice section
- **Intelligent Prediction**: Uses weighted algorithms to predict AP scores
- **Confidence Assessment**: Provides confidence levels based on data completeness

### 2. Manual Score Calculator
- **Real-time Calculation**: Interactive sliders and inputs for immediate feedback
- **Dual Input Methods**: Both slider and numerical input options
- **Score Breakdown**: Detailed component score analysis
- **Composite Scoring**: Automatic calculation of weighted total scores

### 3. Data Persistence & User Integration
- **Save Functionality**: Ability to save prediction results
- **User Authentication**: Integration with backend user system
- **Personal Tracking**: Store and retrieve individual student progress

---

## Technical Architecture

### Frontend Components

#### HTML Structure
- **Responsive Grid Layout**: Adaptive design for desktop and mobile
- **Card-based Interface**: Organized sections for different functionalities
- **Interactive Elements**: Real-time input fields and sliders

#### CSS Styling
- **Modern Dark Theme**: Professional appearance with gradient accents
- **Responsive Design**: Mobile-first approach with breakpoints
- **Interactive Animations**: Hover effects and smooth transitions
- **Typography**: Inter font family for clean readability

#### JavaScript Functionality

##### Core Functions

**`calculatePrediction()`**
- Retrieves practice exam scores and time data
- Calculates weighted composite scores (MCQ: 70%, FRQ: 30%)
- Determines AP score (1-5) based on scaled performance
- Updates confidence levels and strength assessments
- Handles missing data with intelligent estimation

**`updateScores()`**
- Processes manual calculator inputs
- Synchronizes slider and numerical inputs
- Calculates real-time score predictions
- Updates display elements dynamically

**`syncSlider()`**
- Maintains consistency between input methods
- Validates input ranges and constraints
- Triggers score recalculation

##### Scoring Algorithm

```javascript
// Composite Score Calculation
if (mcqCount > 0 && frqCount > 0) {
    compositeScore = (avgMCQ * 0.7) + (practiceFRQ * 5 * 0.3);
} else if (mcqCount > 0) {
    const estimatedFRQ = Math.max(0, Math.min(6, (avgMCQ / 70) * 6));
    compositeScore = (avgMCQ * 0.7) + (estimatedFRQ * 5 * 0.3);
}

// AP Score Mapping
const scaledScore = (compositeScore / 65) * 100;
if (scaledScore >= 90) apScore = 5;
else if (scaledScore >= 81) apScore = 4;
else if (scaledScore >= 42) apScore = 3;
else if (scaledScore >= 30) apScore = 2;
else apScore = 1;
```

#### Configuration Import
```javascript
import { pythonURI } from '{{site.baseurl}}/assets/js/api/config.js';
```

#### User Management
- **Authentication System**: Fetches current user data
- **Data Persistence**: Saves prediction results to backend
- **API Integration**: Communicates with Python backend service

#### Expected Backend Endpoints
- `GET /api/user/current` - Retrieve current user information
- `POST /api/predictions/save` - Save prediction data
- `GET /api/predictions/user/{id}` - Retrieve user's prediction history

---

### Practice Exam Data
```javascript
{
  mcq2018: Number,      // Score out of 70
  mcq2020: Number,      // Score out of 70  
  mcq2021: Number,      // Score out of 70
  practicefrq: Number,  // Score out of 6
  mcq2018Time: Number,  // Minutes spent
  mcq2020Time: Number,  // Minutes spent
  mcq2021Time: Number,  // Minutes spent
  practicefrqTime: Number // Minutes spent
}
```

### Prediction Results
```javascript
{
  userId: String,
  predictedScore: Number,    // 1-5 AP Score
  confidence: String,        // Low/Medium/High
  avgMCQ: Number,           // Average MCQ score
  mcqPercentage: Number,    // MCQ percentage
  frqScore: String,         // FRQ score display
  overallStrength: String,  // Assessment level
  timestamp: Date
}
```

---

#### Story 1: Practice Performance Tracking
**As a** AP CSP student preparing for the exam  
**I want to** input my practice test scores from multiple years  
**So that** I can track my improvement and get an accurate prediction of my AP score

**Acceptance Criteria:**
- I can enter scores for 2018, 2020, and 2021 practice MCQ tests
- I can record my practice FRQ performance
- I can track time spent on each practice section
- The system calculates my predicted AP score based on historical performance
- I receive a confidence level for my prediction

#### Story 2: Real-time Score Calculation
**As a** student taking practice tests  
**I want to** quickly calculate what my AP score would be with different performance levels  
**So that** I can understand what scores I need to achieve my target AP score

**Acceptance Criteria:**
- I can use sliders to adjust MCQ and FRQ scores
- I can enter exact numerical values
- I see my composite score update in real-time
- I see my predicted AP score (1-5) with interpretation
- The interface is responsive and works on my mobile device

#### Story 3: Progress Persistence
**As a** returning student user  
**I want to** save my prediction results  
**So that** I can track my progress over time and review my improvement

**Acceptance Criteria:**
- I can click a "Save Prediction" button
- My prediction data is stored with my user account
- I can access my saved predictions in future sessions
- The system remembers my previous inputs

#### Story 4: Student Assessment Tool
**As a** AP CSP teacher  
**I want to** use the predictor to help students understand their readiness  
**So that** I can provide targeted support and realistic expectations

**Acceptance Criteria:**
- The tool provides clear, understandable predictions
- Confidence levels help me assess reliability of predictions
- Time tracking data helps identify pacing issues
- The interface is professional and appropriate for classroom use

---

### Prerequisites
- Modern web browser with JavaScript enabled
- Backend Python service running
- User authentication system configured

### Installation
1. Include the HTML file in your Jekyll site structure
2. Ensure the CSS and JavaScript are properly loaded
3. Configure the `pythonURI` in your API configuration
4. Set up the required backend endpoints
5. Test the save functionality with your user system

### Configuration Variables
- `pythonURI`: Backend API base URL
- `site.baseurl`: Jekyll site base path
- User authentication endpoints

---

### Frontend Testing
- Test calculation accuracy with known score combinations
- Verify responsive design across devices
- Validate input constraints and error handling
- Test slider/input synchronization

### Backend Integration Testing  
- Verify user authentication flow
- Test data persistence functionality
- Validate API error handling
- Test cross-browser compatibility

### User Experience Testing
- Confirm intuitive navigation
- Validate prediction accuracy against known results
- Test accessibility features
- Verify mobile responsiveness

---

### Potential Features
- **Historical Analysis**: Charts showing score improvement over time
- **Goal Setting**: Target score tracking with progress indicators
- **Study Recommendations**: Personalized study suggestions based on weak areas
- **Comparison Tools**: Anonymous peer comparison features
- **Export Functionality**: PDF reports of prediction results
- **Integration**: LMS integration for classroom deployment

### Technical Improvements
- **Caching**: Local storage for better performance
- **Offline Mode**: Basic functionality without internet
- **Progressive Web App**: Installable mobile experience
- **Advanced Analytics**: More sophisticated prediction algorithms
- **Accessibility**: Enhanced screen reader support
- **Internationalization**: Multi-language support

---

### Regular Updates Needed
- AP scoring criteria updates (if College Board changes standards)
- Practice test data validation
- Backend API compatibility
- Security patches for dependencies

### Performance Monitoring
- Track calculation accuracy
- Monitor user engagement metrics
- Analyze prediction confidence vs. actual results
- Review mobile performance metrics

---

*This documentation serves as a comprehensive guide for developers, educators, and students using the AP Computer Science Principles Exam Predictor application.*
